pglock/core.py

```diff
@@ -0,0 +1,394 @@
+            flag when running as a context manager. Runtime behavior for both the context manager
+            and decorator can be overridden with the ``side_effect`` argument.
+        using (str, default="default"): The database to use.
+        timeout (datetime.timedelta, default=None): Set a timeout when waiting for the lock.
+            This timeout only applies to the lock acquisition statement and not the wrapped code.
+            Only significant when using ``wait=True``. Use ``pglock.timedelta`` as a shortcut to
+            avoid importing Python's timedelta object.
+        side_effect (str, default=pglock.RETURN): Adjust the runtime behavior.
+            ``pglock.RETURN`` will return the acquisition status when using the context manager.
+            ``pglock.RAISE`` will raise a ``django.db.utils.OperationalError`` if the lock cannot
+            be acquired or a timeout happens. ``pglock.SKIP`` will skip decoratored code if the
+            lock cannot be acquired or a timeout happens.
+
+    Returns:
+        bool: When using the default side effect, returns ``True`` if the lock was acquired or
+        ``False`` if not.
+
+    Raises:
+        django.db.utils.OperationalError: When a lock cannot be acquired or a timeout happens
+            when using ``side_effect=pglock.RAISE``.
+        ValueError: If an invalid ``side_effect`` is provided or no lock ID is supplied for
+            the context manager.
+        TypeError: If the lock ID is not a string or int.
+    """  # noqa
+
+    def __init__(
+        self,
+        lock_id=None,
+        *,
+        shared=False,
+        wait=True,
+        using=DEFAULT_DB_ALIAS,
+        timeout=None,
+        side_effect=RETURN,
+    ):
+        """Acquire an advisory lock"""
+        if side_effect not in (RETURN, RAISE, SKIP):
+            raise ValueError(
+                "side_effect must be one of pglock.RETURN, pglock.RAISE, or pglock.SKIP"
+            )
+
+        self.acquire = f'pg{"_try" if not wait else ""}_advisory_lock{"_shared" if shared else ""}'
+        self.release = f'pg_advisory_unlock{"_shared" if shared else ""}'
+        self.lock_id = lock_id
+        self.using = using
+        self.timeout = timeout
+        self.side_effect = side_effect
+        self.wait = wait
+        self.shared = shared
+        self._used_as_context_manager = True
+
+    @property
+    def int_lock_id(self):
+        return _cast_lock_id(self.lock_id)
+
+    def __call__(self, func):
+        self._used_as_context_manager = False
+
+        if not self.lock_id:
+            module = inspect.getmodule(func)
+            self.lock_id = f"{module.__name__}.{func.__name__}"
+
+        @functools.wraps(func)
+        def inner(*args, **kwargs):
+            with self._recreate_cm():
+                if self.acquired or self.side_effect != SKIP:
+                    return func(*args, **kwargs)
```

Comment:
I have yet to see a decorator that works this way in practice. Do you have any examples of libraries that do it? Plus it would have to be a thread local or some other way of making it thread safe.

`mock.patch` supplies the return value as an argument to the decorated function, but that's a pretty special case for test functions that are not imported and called by outside code

I think this is something that can be added later if requested

Reply:
Good point about the thread local nature of it, I had forgotten about that!
Another simple alternative is that we ask the user that the function they decorate accepts an `acquired` kwarg. But no need to change what we have.